persistence-api/src/main/java/io/stargate/db/cdc/datastore/CDCEnabledDataStore.java

```diff
@@ -0,0 +1,119 @@
+import io.stargate.db.datastore.DataStore;
+import io.stargate.db.datastore.ResultSet;
+import io.stargate.db.query.BoundDMLQuery;
+import io.stargate.db.query.BoundQuery;
+import io.stargate.db.query.Query;
+import io.stargate.db.query.TypedValue;
+import io.stargate.db.schema.Schema;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.UnaryOperator;
+
+public class CDCEnabledDataStore implements DataStore {
+  private final DataStore dataStore;
+  private final CDCConfig cdcConfig;
+  private final CDCQueryBuilder cdcQueryBuilder;
+
+  public CDCEnabledDataStore(
+      DataStore dataStore, ConfigStore configStore, ShardManager shardManager) {
+    this(dataStore, configStore, new CDCQueryBuilder(configStore, dataStore, shardManager));
+  }
+
+  @VisibleForTesting
+  public CDCEnabledDataStore(
+      DataStore dataStore, ConfigStore configStore, CDCQueryBuilder cdcQueryBuilder) {
+    this.dataStore = dataStore;
+    this.cdcConfig = CDCConfigLoader.loadConfig(configStore);
+    this.cdcQueryBuilder = cdcQueryBuilder;
+  }
+
+  @Override
+  public TypedValue.Codec valueCodec() {
+    return dataStore.valueCodec();
+  }
+
+  @Override
+  public <B extends BoundQuery> CompletableFuture<Query<B>> prepare(Query<B> query) {
+    return dataStore.prepare(query);
+  }
+
+  @Override
+  public Schema schema() {
+    return dataStore.schema();
+  }
+
+  @Override
+  public boolean isInSchemaAgreement() {
+    return dataStore.isInSchemaAgreement();
+  }
+
+  @Override
+  public void waitForSchemaAgreement() {
+    dataStore.waitForSchemaAgreement();
+  }
+
+  @Override
+  public CompletableFuture<ResultSet> execute(
+      BoundQuery query, UnaryOperator<Parameters> parametersModifier) {
+    if (isMutationTrackedByCDC(query)) {
+      List<BoundQuery> queries = new ArrayList<>();
+      queries.add(query);
+      queries.add(cdcQueryBuilder.toInsert((BoundDMLQuery) query));
+      return dataStore.batch(queries, BatchType.LOGGED, parametersModifier);
+    } else {
+      return dataStore.execute(query, parametersModifier);
+    }
+  }
+
+  @Override
+  public CompletableFuture<ResultSet> batch(
+      Collection<BoundQuery> queries,
+      BatchType batchType,
+      UnaryOperator<Parameters> parametersModifier) {
+    Collection<BoundQuery> cdcInsertQueries = new ArrayList<>();
+    for (BoundQuery query : queries) {
+      if (isMutationTrackedByCDC(query)) {
+        cdcInsertQueries.add(cdcQueryBuilder.toInsert((BoundDMLQuery) query));
+      }
+    }
+    if (cdcInsertQueries.isEmpty()) {
+      return dataStore.batch(queries, batchType, parametersModifier);
+    } else {
+      Collection<BoundQuery> originalAndCDCQueries = new ArrayList<>();
+      originalAndCDCQueries.addAll(queries);
+      originalAndCDCQueries.addAll(cdcInsertQueries);
+      return dataStore.batch(originalAndCDCQueries, BatchType.LOGGED, parametersModifier);
```

Comment:
This might be a surprising result for users whose unlogged batches suddenly turn into logged batches.